datumaro/components/project.py

```diff
@@ -2,486 +2,2389 @@
-                return super().get(id, subset)
+    def save(self, save_dir=None, **kwargs):
+        if save_dir is None and self.readonly:
+            raise ReadonlyDatasetError()
+        self._dataset.save(save_dir, **kwargs)
 
-    def __init__(self, project):
-        super().__init__()
+    @property
+    def readonly(self):
+        return self._readonly or not self.is_bound
+
+    @property
+    def config(self):
+        return self._config
+
+    def __getattr__(self, name):
+        return getattr(self._dataset, name)
+
+    def __setattr__(self, name, value):
+        return setattr(self._dataset, name, value)
 
+    def __iter__(self):
+        yield from self._dataset
+
+    def __len__(self):
+        return len(self._dataset)
+
+    def subsets(self):
+        return self._dataset.subsets()
+
+    def get_subset(self, name):
+        return self._dataset.get_subset(name)
+
+    def categories(self):
+        return self._dataset.categories()
+
+    def get(self, id, subset=None):
+        return self._dataset.get(id, subset)
+
+
+class IgnoreMode(Enum):
+    rewrite = auto()
+    append = auto()
+    remove = auto()
+
+def _update_ignore_file(paths: Union[str, List[str]], repo_root: str,
+        filepath: str, mode: Optional[IgnoreMode] = None):
+    def _make_ignored_path(path):
+        path = osp.join(repo_root, osp.normpath(path))
+        assert path.startswith(repo_root), path
+        return osp.relpath(path, repo_root)
+
+    header = '# The file is autogenerated by Datumaro'
+
+    mode = parse_str_enum_value(mode, IgnoreMode, IgnoreMode.append)
+
+    if isinstance(paths, str):
+        paths = [paths]
+    paths = {osp.join(repo_root, osp.normpath(p)): _make_ignored_path(p)
+        for p in paths}
+
+    openmode = 'r+'
+    if not osp.isfile(filepath):
+        openmode = 'w+' # r+ cannot create, w+ truncates
+    with open(filepath, openmode) as f:
+        lines = []
+        if mode in {IgnoreMode.append, IgnoreMode.remove}:
+            for line in f:
+                lines.append(line.strip())
+            f.seek(0)
+
+        new_lines = []
+        for line in lines:
+            if not line or line.startswith('#'):
+                new_lines.append(line)
+                continue
+
+            line_path = osp.join(repo_root,
+                osp.normpath(line.split('#', maxsplit=1)[0]).lstrip('/'))
+
+            if mode == IgnoreMode.append:
+                if line_path in paths:
+                    paths.pop(line_path)
+                new_lines.append(line)
+            elif mode == IgnoreMode.remove:
+                if line_path in paths:
+                    paths.pop(line_path)
+                else:
+                    new_lines.append(line)
+
+        if mode in { IgnoreMode.rewrite, IgnoreMode.append }:
+            new_lines.extend(paths.values())
+
+        if not new_lines or new_lines[0] != header:
+            f.write(header + '\n')
+        f.writelines('\n'.join(new_lines))
```

Comment:
```suggestion
        for new_line in new_lines:
            print(new_line, file=f)
```

This avoids forming a string with the entire contents of the file, and makes sure that the file ends with a newline.

(Also, I think your original code writes the characters one at a time, because `writeline` accepts an iterable, not a string)

Reply:
> The file argument must be an object with a write(string) method; if it is not present or None, sys.stdout will be used.

This description of the `file` argument makes me think it was designed not as a replacement for `write`. https://docs.python.org/3/library/functions.html#print

Changed to write.

Reply:
> This description of the file argument makes me think it was designed not as a replacement for write

Why? All I see is that it can't be used with binary files.

And that's not the main point of that suggestion anyway, the main point is to use a loop and terminate each loop with a newline.

Reply:
> Why? All I see is that it can't be used with binary files.

I see that this approach can silently start printing to stdout if the method is undefined for some reason instead of failing. I find it fragile.

> And that's not the main point of that suggestion anyway, the main point is to use a loop and terminate each loop with a newline.

With the expected size of file less than 1 kb looping or joining does not matter.

Reply:
> I see that this approach can silently start printing to stdout if the method is undefined for some reason instead of failing. I find it fragile.

I think you have misread the documentation. `stdout` is used if the `file` argument is not specified, not if its value has no `write` method.

> With the expected size of file less than 1 kb looping or joining does not matter.

Fine, but can you at least add a trailing newline?

Reply:
Fixed

Reply:
FWIW, I quite like using `print` for writing files in line-based formats, because it ensures that you don't forget the line terminator. It automatically casting the arguments to `str` and joining them with spaces (or a custom separator) can be useful too.